### PR TITLE
INT-3599/INT-2042: Improve `XPathMessageSplitter`

### PR DESCRIPTION
**Pull request by artembilan**
JIRA: https://jira.spring.io/browse/INT-2042,
https://jira.spring.io/browse/INT-3599
- Expose `outputProperties` for the `XPathMessageSplitter`
- Add `iterator` mode for the `XPathMessageSplitter`

### CODE REVIEW

File: spring-integration-amqp/src/test/java/org/springframework/integration/amqp/config/AmqpChannelParserTests-context.xml [hunk 1]
```diff
@@ -9,7 +9,9 @@
 		http://www.springframework.org/schema/rabbit http://www.springframework.org/schema/rabbit/spring-rabbit.xsd
 		http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans.xsd">
 
-	<amqp:channel id="channelWithInterceptor">
+	<rabbit:queue id="fooQueue" name="my.case.fooQueue" auto-delete="true" durable="false"/>
+
+	<amqp:channel id="channelWithInterceptor" queue-name="my.case.fooQueue">
 		<amqp:interceptors>
```
Comment (garyrussell) [inline spring-integration-amqp/src/test/java/org/springframework/integration/amqp/config/AmqpChannelParserTests-context.xml]: Why is this in this PR?

Comment (artembilan) [inline spring-integration-amqp/src/test/java/org/springframework/integration/amqp/config/AmqpChannelParserTests-context.xml]: Doh. Side-affect of non-commited/rejected changes.
Must be reverted.

File: spring-integration-xml/src/main/resources/org/springframework/integration/xml/config/spring-integration-xml-4.2.xsd [hunk 2]
```diff
@@ -749,6 +749,30 @@
 							<xsd:union memberTypes="xsd:boolean xsd:string" />
 						</xsd:simpleType>
 					</xsd:attribute>
+					<xsd:attribute name="output-properties">
+						<xsd:annotation>
+							<xsd:documentation>
+								A set of output properties that will be used to override any of the same properties
+								in affect for the transformation.
```
Comment (garyrussell) [inline spring-integration-xml/src/main/resources/org/springframework/integration/xml/config/spring-integration-xml-4.2.xsd]: s/affect/effect/

File: spring-integration-xml/src/main/resources/org/springframework/integration/xml/config/spring-integration-xml-4.2.xsd [hunk 3]
```diff
@@ -749,6 +749,30 @@
...
+									<tool:expected-type type="java.util.Properties"/>
+								</tool:annotation>
+							</xsd:appinfo>
+						</xsd:annotation>
+					</xsd:attribute>
+					<xsd:attribute name="iterator" default="true">
+						<xsd:annotation>
+							<xsd:documentation>
+								The iterator mode: 'true' (default) to return an 'java.util.Iterator'
+								for splitting 'payload', 'false to return a 'java.util.List'.
```
Comment (garyrussell) [inline spring-integration-xml/src/main/resources/org/springframework/integration/xml/config/spring-integration-xml-4.2.xsd]: We should explain the benefit - the list contains transformed nodes whereas with the iterator each node is transformed while iterating.

**Comment by artembilan** [comment]
OK. I'll fix those issues. Thank you.
I just worry if you are fine with the whole change.

File: spring-integration-xml/src/main/java/org/springframework/integration/xml/config/XPathMessageSplitterParser.java [hunk 4]
```diff
@@ -47,7 +50,11 @@ protected BeanDefinitionBuilder parseHandler(Element element, ParserContext pars
 		Assert.isTrue(hasChild ^ hasReference, "Exactly one of 'xpath-expression' or 'xpath-expression-ref' is required.");
 		if (hasChild) {
 			BeanDefinition beanDefinition = this.xpathParser.parse((Element) xPathExpressionNodes.item(0), parserContext);
-			builder.addConstructorArgValue(beanDefinition);
+			Map<Integer, ConstructorArgumentValues.ValueHolder> indexedArgumentValues =
+					beanDefinition.getConstructorArgumentValues().getIndexedArgumentValues();
+			for (ConstructorArgumentValues.ValueHolder valueHolder : indexedArgumentValues.values()) {
+				builder.addConstructorArgValue(valueHolder.getValue());
+			}
 		}
```
Comment (garyrussell) [inline spring-integration-xml/src/main/java/org/springframework/integration/xml/config/XPathMessageSplitterParser.java]: This seems a bit dirty - parse and register an `XPathExpressionFactory` but then only use the factory method constructor args.

Perhaps we need to refactor the logic into a utility to add these constructor args to each builder.

**Comment by garyrussell** [comment]
Otherwise it looks ok.

**Comment by artembilan** [comment]
Pushed

**Comment by garyrussell** [comment]
Doc polish here https://github.com/garyrussell/spring-integration/commit/005147c3b41f669cbcf48cc5d2d30a95a23b0655 for your review.

Notice the asciidoctor best practice of starting new sentences on new lines.

**Comment by garyrussell** [comment]
Merged as 0cbba5d43eb06f1b7809fa84a7a75165a3c5045d

**Pull request closed without merging**